### PR TITLE
CE-478: Add tabs to graphics option settings.

### PR DESCRIPTION
**Pull request by serishema**
This adds a tab for each category of graphics options.
press left and right on the dpad to move between tabs.

### CODE REVIEW

File: src/ui/widgets/tabbox.cpp [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  for(int i = 0; i < m_tabs.size(); i++)
+  {
+    m_tabs[i]->update(i == m_selectedTabIndex); 
+  }
+}
+
+void TabBox::draw(ui::Ui& ui, const engine::Presenter& presenter) const
+{
+  int tab_x_offset = 0;
+  for(auto& tab : m_tabs)
```
Comment (stohrendorf) [inline src/ui/widgets/tabbox.cpp:75]:
```suggestion
  for(const auto& [container, content] : m_tabs)
```

File: src/menu/rendersettingsmenustate.h [hunk 2]
```diff
@@ -40,7 +42,8 @@ class RenderSettingsMenuState : public SelectedMenuState
   size_t m_currentListBox = 0;
```
Comment (stohrendorf) [inline src/menu/rendersettingsmenustate.h:42]: `m_currentListBox` is obsolete now, as it resembles the current tab index, please remove it.

File: src/menu/rendersettingsmenustate.cpp [hunk 3]
```diff
@@ -125,8 +123,15 @@ RenderSettingsMenuState::RenderSettingsMenuState(const std::shared_ptr<MenuRingT
...
 
-  auto listBox = std::make_shared<CheckListBox>(/* translators: TR charmap encoding */ _("Effects"));
+  
+  auto tab = std::make_shared<ui::widgets::Tab>(/* translators: TR charmap encoding */ _("Effects"));
+  
+
+  auto listBox = std::make_shared<CheckListBox>();
   m_listBoxes.emplace_back(listBox);
+  m_tabs = std::make_unique<ui::widgets::TabBox>();
+  m_tabs->addTab(tab, std::shared_ptr<ui::widgets::Widget>(listBox));
```
Comment (stohrendorf) [inline src/menu/rendersettingsmenustate.cpp:133]: this is a redundant
```suggestion
  m_tabs->addTab(tab, listBox);
```

File: src/menu/rendersettingsmenustate.cpp [hunk 4]
```diff
@@ -178,8 +183,14 @@ RenderSettingsMenuState::RenderSettingsMenuState(const std::shared_ptr<MenuRingT
...
     });
 
-  listBox = std::make_shared<CheckListBox>(/* translators: TR charmap encoding */ _("Quality"));
+    
+
+  listBox = std::make_shared<CheckListBox>();
   m_listBoxes.emplace_back(listBox);
+  tab = std::make_shared<ui::widgets::Tab>(/* translators: TR charmap encoding */ _("Quality"));
+  
+  m_tabs->addTab(tab, std::shared_ptr<ui::widgets::Widget>(listBox));
```
Comment (stohrendorf) [inline src/menu/rendersettingsmenustate.cpp:192]: this is redundant
```suggestion
  m_tabs->addTab(tab, listBox);
```

File: src/menu/rendersettingsmenustate.cpp [hunk 5]
```diff
@@ -271,8 +282,11 @@ RenderSettingsMenuState::RenderSettingsMenuState(const std::shared_ptr<MenuRingT
       toggle(engine, engine.getEngineConfig()->renderSettings.doubleUiScale);
     });
 
-  listBox = std::make_shared<CheckListBox>(/* translators: TR charmap encoding */ _("Other"));
+  listBox = std::make_shared<CheckListBox>();
   m_listBoxes.emplace_back(listBox);
+  tab = std::make_shared<ui::widgets::Tab>(/* translators: TR charmap encoding */ _("Other"));
+  m_tabs->addTab(tab, std::shared_ptr<ui::widgets::Widget>(listBox));
```
Comment (stohrendorf) [inline src/menu/rendersettingsmenustate.cpp:288]:
```suggestion
  m_tabs->addTab(tab, listBox);
```

File: src/menu/rendersettingsmenustate.cpp [hunk 6]
```diff
@@ -271,8 +282,11 @@ RenderSettingsMenuState::RenderSettingsMenuState(const std::shared_ptr<MenuRingT
       toggle(engine, engine.getEngineConfig()->renderSettings.doubleUiScale);
     });
 
-  listBox = std::make_shared<CheckListBox>(/* translators: TR charmap encoding */ _("Other"));
+  listBox = std::make_shared<CheckListBox>();
```
Comment (stohrendorf) [inline src/menu/rendersettingsmenustate.cpp:285]:
```suggestion
  listBox = gslu::make_nn_shared<CheckListBox>();
```

File: src/menu/rendersettingsmenustate.cpp [hunk 7]
```diff
@@ -178,8 +183,14 @@ RenderSettingsMenuState::RenderSettingsMenuState(const std::shared_ptr<MenuRingT
       toggle(engine, engine.getEngineConfig()->renderSettings.velvia);
     });
 
-  listBox = std::make_shared<CheckListBox>(/* translators: TR charmap encoding */ _("Quality"));
+    
+
+  listBox = std::make_shared<CheckListBox>();
```
Comment (stohrendorf) [inline src/menu/rendersettingsmenustate.cpp:188]:
```suggestion
  listBox = gslu::make_nn_shared<CheckListBox>();
```

File: src/menu/rendersettingsmenustate.cpp [hunk 8]
```diff
@@ -178,8 +183,14 @@ RenderSettingsMenuState::RenderSettingsMenuState(const std::shared_ptr<MenuRingT
...
   m_listBoxes.emplace_back(listBox);
+  tab = std::make_shared<ui::widgets::Tab>(/* translators: TR charmap encoding */ _("Quality"));
```
Comment (stohrendorf) [inline src/menu/rendersettingsmenustate.cpp:190]:
```suggestion
  tab = gslu::make_nn_shared<ui::widgets::Tab>(/* translators: TR charmap encoding */ _("Quality"));
```

File: src/menu/rendersettingsmenustate.cpp [hunk 9]
```diff
@@ -125,8 +123,15 @@ RenderSettingsMenuState::RenderSettingsMenuState(const std::shared_ptr<MenuRingT
     engine.applySettings();
   };
 
-  auto listBox = std::make_shared<CheckListBox>(/* translators: TR charmap encoding */ _("Effects"));
+  
+  auto tab = std::make_shared<ui::widgets::Tab>(/* translators: TR charmap encoding */ _("Effects"));
```
Comment (stohrendorf) [inline src/menu/rendersettingsmenustate.cpp:127]:
```suggestion
  auto tab = gslu::make_nn_shared<ui::widgets::Tab>(/* translators: TR charmap encoding */ _("Effects"));
```

File: src/menu/rendersettingsmenustate.cpp [hunk 10]
```diff
@@ -369,54 +383,49 @@ std::unique_ptr<MenuState>
...
   }
   if(world.getPresenter().getInputHandler().getInputState().xMovement.justChangedTo(hid::AxisMovement::Left))
   {
     if(m_currentListBox > 0)
+    {
       --m_currentListBox;
+      m_tabs->prevTab();
+    }
   }
   else if(world.getPresenter().getInputHandler().getInputState().xMovement.justChangedTo(hid::AxisMovement::Right))
```
Comment (stohrendorf) [inline src/menu/rendersettingsmenustate.cpp:422]: please use step left/right for moving between the tabs. once the slider widget will be added, plain left/right will be moved to change the slider value. also, you may want to change the input mapping widget to use the tab widget instead, which already uses step left/right to switch between input mappings.

File: src/menu/rendersettingsmenustate.cpp [hunk 11]
```diff
@@ -125,8 +123,15 @@ RenderSettingsMenuState::RenderSettingsMenuState(const std::shared_ptr<MenuRingT
     engine.applySettings();
   };
 
-  auto listBox = std::make_shared<CheckListBox>(/* translators: TR charmap encoding */ _("Effects"));
+  
+  auto tab = std::make_shared<ui::widgets::Tab>(/* translators: TR charmap encoding */ _("Effects"));
+  
+
+  auto listBox = std::make_shared<CheckListBox>();
```
Comment (stohrendorf) [inline src/menu/rendersettingsmenustate.cpp:130]:
```suggestion
  auto listBox = gslu::make_nn_shared<CheckListBox>();
```

File: src/ui/widgets/tabbox.cpp [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+{
+
+constexpr int TitleOffset = 10;
+constexpr int InnerMargin = 10;
+constexpr int WidgetOffsetTop = ui::FontHeight / 2 + InnerMargin + ui::OutlineBorderWidth;
+constexpr int TotalVerticalMargin = WidgetOffsetTop + InnerMargin + 2 * ui::OutlineBorderWidth;
+
+
+Tab::Tab(const std::string& title)
+    : m_title(std::make_unique<Checkbox>(title))
```
Comment (stohrendorf) [inline src/ui/widgets/tabbox.cpp:17]: I think the visual layout is too complex. maybe something like this works better - dimming inactive tabs, add a lot of spacing between headers, and a smaller rectangle next to the active tab. other ideas are welcome, too.

grafik

(positioning isn't accurate, it's more about the colors I wanted to show.)

**Review by stohrendorf — CHANGES_REQUESTED** [review]
these are just change requests about the most problematic things, once these are fixed, I'll check again. one notable thing is that the menu is now centered on my screen, and not with an offset from the bottom like the others, which is, if I remember correctly, 90px.

File: src/menu/rendersettingsmenustate.cpp (hunk 6 above)
Comment (serishema) [reply src/menu/rendersettingsmenustate.cpp:285]: gslu::make_nn_shared was deleted in [commit 6f5e13b3e6b1355d10db34e813c1129a391fa90b ](https://github.com/stohrendorf/CroftEngine/commit/6f5e13b3e6b1355d10db34e813c1129a391fa90b)
What should I do instead?

File: src/menu/rendersettingsmenustate.cpp [hunk 13]
```diff
@@ -125,8 +123,15 @@ RenderSettingsMenuState::RenderSettingsMenuState(const std::shared_ptr<MenuRingT
...
   };
 
-  auto listBox = std::make_shared<CheckListBox>(/* translators: TR charmap encoding */ _("Effects"));
+  
+  auto tab = gsl::make_shared<ui::widgets::Tab>(/* translators: TR charmap encoding */ _("Effects"));
+  
+
+  auto listBox = gsl::make_shared<CheckListBox>();
   m_listBoxes.emplace_back(listBox);
+  m_tabs = std::make_unique<ui::widgets::TabBox>();
```
Comment (stohrendorf) [inline src/menu/rendersettingsmenustate.cpp:132]: Please move this to the constructor initializer list.

File: src/menu/rendersettingsmenustate.cpp [hunk 14]
```diff
@@ -369,54 +383,31 @@ std::unique_ptr<MenuState>
...
-    for(const auto& listBox : m_listBoxes)
-    {
-      const auto height = listBox->getSize().y;
-      listBox->setSize({maxW, height});
-      listBox->setPosition({(ui.getSize().x - maxW) / 2, y});
-      y += height + Separation;
-    }
-  }
+  m_tabs->fitToContent();
+  m_tabs->setPosition({(ui.getSize().x) / 2 - (m_tabs->getSize().x / 2), ui.getSize().y -m_tabs->getSize().y - 90});
```
Comment (stohrendorf) [inline src/menu/rendersettingsmenustate.cpp:387]:
```suggestion
  m_tabs->setPosition({(ui.getSize().x - m_tabs->getSize().x / 2), ui.getSize().y - m_tabs->getSize().y - 90});
```

File: src/menu/rendersettingsmenustate.cpp [hunk 15]
```diff
@@ -369,54 +383,31 @@ std::unique_ptr<MenuState>
...
 
   for(size_t i = 0; i < m_listBoxes.size(); ++i)
   {
     const auto& listBox = m_listBoxes[i];
-    listBox->update(i == m_currentListBox);
-    listBox->draw(ui, world.getPresenter());
+    listBox->update(i == m_tabs->getSelectedTab());
```
Comment (stohrendorf) [inline src/menu/rendersettingsmenustate.cpp:392]:
```suggestion
    m_listBoxes[i]->update(i == m_tabs->getSelectedTab());
```
the temporary variable `const auto& listBox = m_listBoxes[i];` can then be removed

File: src/menu/rendersettingsmenustate.h [hunk 16]
```diff
@@ -37,10 +39,10 @@ class RenderSettingsMenuState : public SelectedMenuState
 private:
   class CheckListBox;
   std::vector<std::shared_ptr<CheckListBox>> m_listBoxes{};
-  size_t m_currentListBox = 0;
   std::unique_ptr<MenuState> m_previous;
   std::shared_ptr<ui::widgets::Checkbox> m_anisotropyCheckbox;
-
+  std::unique_ptr<ui::widgets::TabBox> m_tabs;
```
Comment (stohrendorf) [inline src/menu/rendersettingsmenustate.h:44]:
```suggestion
  gslu::nn_unique<ui::widgets::TabBox> m_tabs;
```

File: src/ui/widgets/tabbox.cpp [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+  for(int i = 0; i < m_tabs.size(); i++)
+  {
+    const auto& [container, content] = m_tabs[i];
+    container->update(i == m_selectedTabIndex); 
+  }
+}
+
+void TabBox::draw(ui::Ui& ui, const engine::Presenter& presenter) const
+{
+  int tab_x_offset = 0;
```
Comment (stohrendorf) [inline src/ui/widgets/tabbox.cpp:82]:
```suggestion
  int tabOffsetX = 0;
```

File: src/ui/widgets/tabbox.cpp [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+  }
+  const auto& [container, content] = m_tabs[m_selectedTabIndex];
+  content->fitToContent();
+  content->setPosition(m_position + glm::ivec2{0, TotalVerticalMargin});
+  content->draw(ui, presenter);
+}
+
+void TabBox::addTab(const gslu::nn_shared<Tab>& tab, const gslu::nn_shared<Widget>& content)
+{
+  m_tabs.emplace_back(std::make_tuple(tab, content));
```
Comment (stohrendorf) [inline src/ui/widgets/tabbox.cpp:111]:
```suggestion
  m_tabs.emplace_back(tab, content);
```

**Review by stohrendorf — CHANGES_REQUESTED** [review]
looks better now

File: src/menu/rendersettingsmenustate.cpp (hunk 6 above)
Comment (stohrendorf) [reply src/menu/rendersettingsmenustate.cpp]: Ah, sorry, I meant `gsl::make_shared`.

File: src/menu/rendersettingsmenustate.cpp (hunk 7 above)
Comment (stohrendorf) [reply src/menu/rendersettingsmenustate.cpp]: Same here, `gsl::make_shared`

File: src/menu/rendersettingsmenustate.cpp (hunk 8 above)
Comment (stohrendorf) [reply src/menu/rendersettingsmenustate.cpp]: And, again, `gsl::make_shared`

File: src/menu/rendersettingsmenustate.cpp (hunk 9 above)
Comment (stohrendorf) [reply src/menu/rendersettingsmenustate.cpp]: `gsl::make_shared`

File: src/menu/rendersettingsmenustate.cpp (hunk 11 above)
Comment (stohrendorf) [reply src/menu/rendersettingsmenustate.cpp]: `gsl::make_shared`

File: src/ui/widgets/tabbox.cpp (hunk 12 above)
Comment (stohrendorf) [reply src/ui/widgets/tabbox.cpp:17]: Maybe just make the header checkboxes borderless to remove some clutter.

Comment (serishema) [reply src/ui/widgets/tabbox.cpp:17]: I'll adpot that suggestion, i'm having difficulty drawing dimmed text.

Comment (stohrendorf) [reply src/ui/widgets/tabbox.cpp]: looks good what you did.

File: src/menu/rendersettingsmenustate.cpp (hunk 14 above)
Comment (serishema) [reply src/menu/rendersettingsmenustate.cpp]: This results in an X position of 815 which is half off the screen. 
I have used 
`{ui.getSize().x - m_container->getSize().x) / 2,ui.getSize().y - m_tabs->getSize().y - 90`}
the same as in audio settings.

Comment (stohrendorf) [reply src/menu/rendersettingsmenustate.cpp]: yes, sorry, misplaced `)`, that should have gone after `getSize().x` in the suggestion

File: src/menu/rendersettingsmenustate.h [hunk 19]
```diff
@@ -1,11 +1,13 @@
 #pragma once
 
+#include "ui/widgets/tabbox.h"
```
Comment (stohrendorf) [inline src/menu/rendersettingsmenustate.h:3]: The issue with the compiler complaining about deleting an incomplete type came from an implicit destructor implementation. The `RenderSettingsMenuState` class doesn't explicitly declare a destructor, which means the destructor is synthetically generated as an inline destructor. By including the missing incomplete type, you solved that, but in turn made the `TabBox` an implicit, public dependency of this header, which is not necessary, as it's an implementation detail. The better approach would be to just declare a destructor in the header (`~RenderSettingsMenuState() override;`) and then force the implementation to be in the `.cpp` by instantiating the default implementation there (`RenderSettingsMenuState::~RenderSettingsMenuState() = default;`).

**Review by stohrendorf — CHANGES_REQUESTED** [review]

**Comment by stohrendorf** [comment]
Looks good. One last thing: when scrolling through the checkboxes of a tab, make it so that it wraps around within the list instead of switching to the next/previous tab. Once that's done, please squash all commits into a single one, and rebase onto the master branch. If possible, also let clang-format reformat the code. Good work.

**Comment by serishema** [comment]
Thanks, i've made up/down wrap in the list, squashed and rebased this against orgin/master.

**Pull request merged**